sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/SafeAddTests.java

```diff
@@ -0,0 +1,79 @@
+package org.dbsp.sqlCompiler.compiler.sql.simple;
+import org.dbsp.sqlCompiler.compiler.sql.SqlIoTest;
+import org.junit.Ignore;
+import org.junit.Test;
+
+public class SafeAddTests extends SqlIoTest {
+    @Test
+    public void testSafeAdd() {
+        this.qs(
+                "SELECT SAFE_ADD(2, 3) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                "5\n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT SAFE_ADD(9223372036854775807, 2) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                " \n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT SAFE_ADD(-9223372036854775807, -2) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                " \n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT SAFE_ADD(9.9, 1.0) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                "10.9\n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT SAFE_ADD(9.9, 1) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                "10.9\n" +
+                "(1 row)\n" +
+                "\n" +
+                "SELECT SAFE_ADD(CAST(125 AS TINYINT), CAST(2 AS TINYINT)) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                " 127\n" +
+                "(1 row)\n" +
+                "\n"
+        );
+    }
+
+    @Test
+    @Ignore
+    public void testSafeAddWithCasts() {
+        this.q(
+        "SELECT SAFE_ADD(CAST(32767 AS SMALLINT), CAST(2 AS SMALLINT)) as 'result';\n" +
+                " result \n" +
+                "--------\n" +
+                " \n"
+        );
+    }
+
+    @Test
+    @Ignore
+    public void testSafeAddWithNestedSelectCast() {
```

Comment:
this test is actually not a good test given the definition of safe_add: it will always convert its arguments to int64 before doing the addition.  So the result is correct, but this shows that the function is very misleading.

Reply:
Ah hmm. 
One more thing about this test is that, it fails to compile in rust as Rust expects: `Option<T>` but gets `Option<Option<T>>`.  It should get flattened, I think. 

Error: 
```rust
/Library/Java/JavaVirtualMachines/temurin-19.jdk/Contents/Home/bin/java -ea -Didea.test.cyclic.buffer.size=1048576 -javaagent:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=49765:/Applications/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8 -Dsun.stdout.encoding=UTF-8 -Dsun.stderr.encoding=UTF-8 -classpath /Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/junit/lib/junit5-rt.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/junit/lib/junit-rt.jar:/Users/abhizer/feldera/feldera/sql-to-dbsp-compiler/SQL-compiler/target/test-classes:/Users/abhizer/feldera/feldera/sql-to-dbsp-compiler/SQL-compiler/target/classes:/Users/abhizer/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-csv/2.12.2/jackson-dataformat-csv-2.12.2.jar:/Users/abhizer/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.12.2/jackson-databind-2.12.2.jar:/Users/abhizer/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.12.2/jackson-annotations-2.12.2.jar:/Users/abhizer/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.2/jackson-core-2.12.2.jar:/Users/abhizer/.m2/repository/com/google/guava/guava/32.0.1-jre/guava-32.0.1-jre.jar:/Users/abhizer/.m2/repository/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar:/Users/abhizer/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar:/Users/abhizer/.m2/repository/org/checkerframework/checker-qual/3.33.0/checker-qual-3.33.0.jar:/Users/abhizer/.m2/repository/com/google/errorprone/error_prone_annotations/2.18.0/error_prone_annotations-2.18.0.jar:/Users/abhizer/.m2/repository/com/google/j2objc/j2objc-annotations/2.8/j2objc-annotations-2.8.jar:/Users/abhizer/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/Users/abhizer/.m2/repository/org/apache/commons/commons-text/1.10.0/commons-text-1.10.0.jar:/Users/abhizer/.m2/repository/org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.jar:/Users/abhizer/.m2/repository/org/slf4j/slf4j-simple/2.0.5/slf4j-simple-2.0.5.jar:/Users/abhizer/.m2/repository/org/slf4j/slf4j-api/2.0.5/slf4j-api-2.0.5.jar:/Users/abhizer/.m2/repository/org/apache/calcite/calcite-core/1.36.0/calcite-core-1.36.0.jar:/Users/abhizer/.m2/repository/org/apache/calcite/calcite-linq4j/1.36.0/calcite-linq4j-1.36.0.jar:/Users/abhizer/.m2/repository/org/locationtech/jts/jts-core/1.19.0/jts-core-1.19.0.jar:/Users/abhizer/.m2/repository/org/locationtech/jts/io/jts-io-common/1.19.0/jts-io-common-1.19.0.jar:/Users/abhizer/.m2/repository/com/googlecode/json-simple/json-simple/1.1.1/json-simple-1.1.1.jar:/Users/abhizer/.m2/repository/org/locationtech/proj4j/proj4j/1.2.2/proj4j-1.2.2.jar:/Users/abhizer/.m2/repository/org/apache/calcite/avatica/avatica-core/1.23.0/avatica-core-1.23.0.jar:/Users/abhizer/.m2/repository/org/apache/calcite/avatica/avatica-metrics/1.23.0/avatica-metrics-1.23.0.jar:/Users/abhizer/.m2/repository/com/google/protobuf/protobuf-java/3.21.9/protobuf-java-3.21.9.jar:/Users/abhizer/.m2/repository/org/apache/httpcomponents/client5/httpclient5/5.1.3/httpclient5-5.1.3.jar:/Users/abhizer/.m2/repository/org/apache/httpcomponents/core5/httpcore5-h2/5.1.3/httpcore5-h2-5.1.3.jar:/Users/abhizer/.m2/repository/org/apache/httpcomponents/core5/httpcore5/5.1.3/httpcore5-5.1.3.jar:/Users/abhizer/.m2/repository/org/apiguardian/apiguardian-api/1.1.2/apiguardian-api-1.1.2.jar:/Users/abhizer/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-yaml/2.15.0/jackson-dataformat-yaml-2.15.0.jar:/Users/abhizer/.m2/repository/org/yaml/snakeyaml/2.0/snakeyaml-2.0.jar:/Users/abhizer/.m2/repository/com/google/uzaygezen/uzaygezen-core/0.2/uzaygezen-core-0.2.jar:/Users/abhizer/.m2/repository/com/jayway/jsonpath/json-path/2.7.0/json-path-2.7.0.jar:/Users/abhizer/.m2/repository/net/minidev/json-smart/2.4.7/json-smart-2.4.7.jar:/Users/abhizer/.m2/repository/net/minidev/accessors-smart/2.4.7/accessors-smart-2.4.7.jar:/Users/abhizer/.m2/repository/org/ow2/asm/asm/9.1/asm-9.1.jar:/Users/abhizer/.m2/repository/com/yahoo/datasketches/sketches-core/0.9.0/sketches-core-0.9.0.jar:/Users/abhizer/.m2/repository/com/yahoo/datasketches/memory/0.9.0/memory-0.9.0.jar:/Users/abhizer/.m2/repository/commons-codec/commons-codec/1.13/commons-codec-1.13.jar:/Users/abhizer/.m2/repository/net/hydromatic/aggdesigner-algorithm/6.0/aggdesigner-algorithm-6.0.jar:/Users/abhizer/.m2/repository/commons-lang/commons-lang/2.4/commons-lang-2.4.jar:/Users/abhizer/.m2/repository/commons-logging/commons-logging/1.1.3/commons-logging-1.1.3.jar:/Users/abhizer/.m2/repository/org/apache/commons/commons-dbcp2/2.9.0/commons-dbcp2-2.9.0.jar:/Users/abhizer/.m2/repository/org/apache/commons/commons-pool2/2.10.0/commons-pool2-2.10.0.jar:/Users/abhizer/.m2/repository/org/apache/commons/commons-math3/3.6.1/commons-math3-3.6.1.jar:/Users/abhizer/.m2/repository/commons-io/commons-io/2.11.0/commons-io-2.11.0.jar:/Users/abhizer/.m2/repository/org/codehaus/janino/commons-compiler/3.1.9/commons-compiler-3.1.9.jar:/Users/abhizer/.m2/repository/org/codehaus/janino/janino/3.1.9/janino-3.1.9.jar:/Users/abhizer/.m2/repository/org/apache/calcite/calcite-babel/1.36.0/calcite-babel-1.36.0.jar:/Users/abhizer/.m2/repository/org/apache/calcite/calcite-server/1.36.0/calcite-server-1.36.0.jar:/Users/abhizer/.m2/repository/junit/junit/4.13.2/junit-4.13.2.jar:/Users/abhizer/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/Users/abhizer/.m2/repository/org/xerial/sqlite-jdbc/3.42.0.0/sqlite-jdbc-3.42.0.0.jar:/Users/abhizer/.m2/repository/com/beust/jcommander/1.82/jcommander-1.82.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 org.dbsp.sqlCompiler.compiler.sql.simple.SafeAddTests,testSafeAddWithNestedSelectCast
error[E0308]: mismatched types
   --> src/lib.rs:271:25
    |
271 |             Tuple1::new(safe_add_i16N_i16N(t.1, t.2))
    |             ----------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<i16>`, found `Option<Option<i16>>`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected enum `std::option::Option<i16>`
               found enum `std::option::Option<std::option::Option<i16>>`
note: associated function defined here
   --> src/lib.rs:88:1
    |
88  | / declare_tuples! {
89  | |     Tuple1<T0>,
    | |            --
90  | |     Tuple2<T0, T1>,
91  | |     Tuple3<T0, T1, T2>,
92  | | }
    | |_^
    = note: this error originates in the macro `declare_tuples` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> src/lib.rs:478:25
    |
478 |             Tuple1::new(safe_add_i16N_i16N(t.1, t.2))
    |             ----------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<i16>`, found `Option<Option<i16>>`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected enum `std::option::Option<i16>`
               found enum `std::option::Option<std::option::Option<i16>>`
note: associated function defined here
   --> src/lib.rs:88:1
    |
88  | / declare_tuples! {
89  | |     Tuple1<T0>,
    | |            --
90  | |     Tuple2<T0, T1>,
91  | |     Tuple3<T0, T1, T2>,
92  | | }
    | |_^
    = note: this error originates in the macro `declare_tuples` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `temp` (lib test) due to 2 previous errors

java.lang.RuntimeException: Process failed with exit code 101

	at org.dbsp.util.Utilities.runProcess(Utilities.java:202)
	at org.dbsp.util.Utilities.runProcess(Utilities.java:207)
	at org.dbsp.util.Utilities.compile(Utilities.java:221)
	at org.dbsp.util.Utilities.compileAndTestRust(Utilities.java:228)
	at org.dbsp.sqlCompiler.compiler.sql.BaseSQLTests.runAllTests(BaseSQLTests.java:134)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)


Process finished with exit code 255

```